### ISSUE TITLE
[dev] Commiting 19x more memory than is requested 

### ISSUE THREAD

**Issue by mjp41** [jemalloc/jemalloc#446]
I observed je_malloc committing way too much memory on Windows 10. This can be duplicated with a simple loop calling 
```C
char * x = je_malloc(0x10000);
```

If you monitor the calls to Virtual Alloc you observe the following behaviour
```
Call je_malloc requesting: 0x10000
    VirtualAlloc requesting: 0x11000
      VirtualAlloc returned: 0xbe85fe0000
    VirtualAlloc requesting: 0x90000
      VirtualAlloc returned: 0xbe863d0000
    VirtualAlloc requesting: 0x90000
      VirtualAlloc returned: 0xbe86460000
  je_malloc returned: 0xbe85fe0c80
```
After the first call, every subsequent call has this pattern. Allocating a block for the allocation, and then two nodes in the Radix tree.

This results in Virtual Alloc commiting 19x more memory than is requested, and completely killing the swap file. 

I was compiling against the static library Release using Visual Studio 2015: 
```
Microsoft (R) C/C++ Optimizing Compiler Version 19.00.23918 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```

We have seen similar behaviour on Debian Sid. 

This does not happen with 4.2.1.

**Comment by jasone** [comment]
I'm assuming this report is based on 19c9a3e828ed46f1576521c264640e60bd0cb01f.  If not, please update with further info.

**Comment by mjp41** [comment]
It is based on 19c9a3e.

**Comment by ogaun** [comment]
Hi, I'm experiencing the same problem here under Windows 10 with VS2015. 
After some investigations, I found out that the current 19c9a3e828ed46f1576521c264640e60bd0cb01f  calls `rtree_new` with 0x34 bits, creating levels 0x10/0x10/0x10/0x4, whereas v4.2.1 calls `rtree_new`with 0x30 bits with levels 0x10/0x10/0x10/0x0.
This change is due to the fact that 19c9a3e828ed46f1576521c264640e60bd0cb01f  always uses LG_PAGE to compute the number of significant bits, whereas v4.2.1 uses Window's allocation granularity (`opt_lg_chunk`)  (`chunk.c:786`) :
```
	if (rtree_new(&chunks_rtree, (unsigned)((ZU(1) << (LG_SIZEOF_PTR+3)) -
	    opt_lg_chunk), chunks_rtree_node_alloc, NULL))
		return (true);
```
While this may be a regression for Windows, I'd tend to think that it does not explain the massive overallocation. After additional debugging I think there is a mistake when allocating rtree children pages (rtree.c):
```
rtree_elm_t *
rtree_child_read_hard(tsdn_t *tsdn, rtree_t *rtree, rtree_elm_t *elm,
    unsigned level)
{
	return (rtree_node_init(tsdn, rtree, level, &elm->child));
}
```
AFAIU, `level` is the level of the node we want to create a child for, so the child is a `level+1`-level node, so I think it should read:
```
rtree_elm_t *
rtree_child_read_hard(tsdn_t *tsdn, rtree_t *rtree, rtree_elm_t *elm,
    unsigned level)
{
	return (rtree_node_init(tsdn, rtree, level+1, &elm->child));
}
```
This way, leaf nodes (when the last level of the rtree is 4 bits) are much smaller than internal nodes as expected. However, even with this fix, I experience a final VM size (Commit size) of about 11 Gb after a stress test on an application, as compared to 1.5 Gb with v2.4.1 (and 215 Gb with 19c9a3e828ed46f1576521c264640e60bd0cb01f unmodified (!)).
hth :)

**Comment by jasone** [comment]
I apparently didn't read everything above closely during triage, and rediscovered the ```rtree_child_read_hard()``` bug a few days ago (fixed in 68e14c988460907c4b135feb3eb5fccd28953feb).  Thank you @mjp41 for the report, and @ogaun for the fix.  If only I'd taken advantage of your efforts instead of spending a late night replicating them!

See [this comment](https://github.com/jemalloc/jemalloc/issues/393#issuecomment-256861194) on #393 for additional info on memory usage issues specific to Windows and the dev branch.
